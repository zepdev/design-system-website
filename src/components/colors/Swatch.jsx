import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Button from '../ui/Button'
import CopyIcon from '../icons/CopyIcon'

class Swatch extends Component {
  state = {
    isTextCopied: false,
    copyError: false,
  }

  handleCopy = color => {
    if (document.queryCommandSupported('copy')) {
      let dummy = document.createElement('textarea')
      document.body.appendChild(dummy)
      dummy.value = color
      dummy.select()
      document.execCommand('copy')
      document.body.removeChild(dummy)
      this.setState({ isTextCopied: true })
      setTimeout(() => {
        this.setState({ isTextCopied: false })
      }, 2000)
      console.log(document.queryCommandSupported('copy'))
    } else {
      this.setState({ copyError: true })
      setTimeout(() => {
        this.setState({ copyError: false })
      }, 2000)
      console.log(document.queryCommandSupported('copy'))
    }
  }

  render() {
    const { title, color } = this.props
    const { isTextCopied, copyError } = this.state
    return (
      <div className="zds-swatch--container">
        <Button
          className="zds-swatch"
          style={{ background: color }}
          onClick={() => this.handleCopy(color)}
        >
          <span className="zds-swatch--backdrop" />
          <span className="zds-swatch--imageButton">
            {!isTextCopied && <CopyIcon className="zds-swatch--icon" />}
            <span className="zds-swatch--imageTitle" data-testid="swatch-text">
              {isTextCopied ? 'Copied!' : copyError ? 'Error!' : color}
            </span>
          </span>
        </Button>
        <p className="zds-swatch--title">{title}</p>
      </div>
    )
  }
}

Swatch.propTypes = {
  color: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
}

export default Swatch
