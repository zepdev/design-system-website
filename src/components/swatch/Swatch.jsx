import React, { useState } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import withStyles from 'react-jss'
import ButtonBase from '../button/ButtonBase'
import CopyIcon from '../icons/CopyIcon'

const styles = theme => ({
  root: {
    display: 'flex',
    flexDirection: 'column',
    flexBasis: '48%',
    marginBottom: `${ theme.spacing.component.xxl.rem }rem`,
  },
  button: {
    height: '8rem',
    width: '100%',
    marginBottom: theme.spacing.component.s.px,
    position: 'relative',
    cursor: 'pointer',
    border: `1px solid ${ theme.colors.gray.grayMid.hex }`,
    '&:hover, &:focus': {
      zIndex: 1,
      '& $backdrop': {
        opacity: 0.4,
      },
      '& $backdropContent': {
        opacity: 1,
      },
    },
  },
  [`@media (min-width: ${ theme.breakpoints.s })`]: {
    root: {
      flexBasis: '22%',
    },
    button: {
      height: '8rem',
    },
  },
  [`@media (min-width: ${ theme.breakpoints.m })`]: {
    root: {
      marginRight: `${ theme.spacing.component.m.rem }rem`,
    },
  },
  [`@media (min-width: ${ theme.breakpoints.xxl })`]: {
    root: {
      marginRight: `${ theme.spacing.component.m.rem }rem`,
      flexBasis: '24%',
    },
    button: {
      height: '16rem',
    },
  },
  backdrop: {
    position: 'absolute',
    left: 0,
    right: 0,
    top: 0,
    bottom: 0,
    backgroundColor: theme.colors.gray.grayDark.hex,
    opacity: 0,
    transition: 0.5,
  },
  backdropContent: {
    position: 'absolute',
    left: 0,
    right: 0,
    top: 0,
    bottom: 0,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    opacity: 0,
    color: theme.colors.gray.white.hex,
  },
  icon: {
    color: theme.colors.gray.white.hex,
    display: 'block',
  },
  buttonText: {
    position: 'relative,',
    paddingLeft: 5,
    display: 'block',
    textTransform: 'uppercase',
  },
  text: {
    margin: 0,
    color: theme.colors.gray.grayLight.hex,
    textTransform: 'uppercase',
  },
  name: {
    marginTop: 0,
    fontWeight: 500,
    marginBottom: theme.spacing.component.s.px,
  },
})

function Swatch({ color, classes }) {
  const [isTextCopied, setTextCopied] = useState(false)
  const [copyError, setCopyError] = useState(false)

  const handleCopy = color => {
    if (document.queryCommandSupported('copy')) {
      let dummy = document.createElement('textarea')
      document.body.appendChild(dummy)
      dummy.value = color
      dummy.select()
      document.execCommand('copy')
      document.body.removeChild(dummy)
      setTextCopied(true)
      setTimeout(() => {
        setTextCopied(false)
      }, 2000)
    } else {
      setCopyError(true)
      setTimeout(() => {
        setCopyError(false)
      }, 2000)
    }
  }

  return (
    <div className={classes.root}>
      <ButtonBase
        className={classes.button}
        style={{ background: color.hex }}
        onClick={() => handleCopy(color.hex)}
        data-testid="swatchButton"
      >
        <span className={classes.backdrop} />
        <span className={classes.backdropContent}>
          {!isTextCopied && (
            <CopyIcon className={classes.icon} ariaLabel={`copy hex:${ color.hex }`} />
          )}
          <span
            className={classnames(classes.buttonText, 'zep-typo--normal-4')}
            data-testid="swatchText"
          >
            {isTextCopied ? 'Copied!' : copyError ? 'Error!' : color.hex}
          </span>
        </span>
      </ButtonBase>
      <p className={classnames(classes.name, 'zep-typo--normal-2')}>{color.name}</p>
      <p className={classnames(classes.text, 'zep-typo--normal-1')}>{`HEX: ${ color.hex }`}</p>
      {color.rgb && (
        <p className={classnames(classes.text, 'zep-typo--normal-1')}>{`RGB: ${ color.rgb }`}</p>
      )}
      {color.hsb && (
        <p className={classnames(classes.text, 'zep-typo--normal-1')}>{`HSB: ${ color.hsb }`}</p>
      )}
    </div>
  )
}

Swatch.propTypes = {
  classes: PropTypes.object.isRequired,
  color: PropTypes.object.isRequired,
}

export default withStyles(styles)(Swatch)
