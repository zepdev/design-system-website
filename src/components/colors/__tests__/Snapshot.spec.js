import React from 'react'
import { render } from 'react-testing-library'
import Swatch from '../Swatch'

describe('Swatch', () => {
  it('renders correctly', () => {
    const mockProps = { title: 'title', color: '#fff' }
    const component = render(<Swatch {...mockProps} />)
    expect(component).toMatchSnapshot()
  })
})
