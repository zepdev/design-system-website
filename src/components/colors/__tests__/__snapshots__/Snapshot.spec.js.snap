// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swatch renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="zds-swatch--container"
      >
        <div
          class="zds-swatch"
          style="background: rgb(255, 255, 255);"
        />
        <p
          class="zds-swatch--title"
        >
          title
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="zds-swatch--container"
    >
      <div
        class="zds-swatch"
        style="background: rgb(255, 255, 255);"
      />
      <p
        class="zds-swatch--title"
      >
        title
      </p>
    </div>
  </div>,
  "debug": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllBySelectText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getBySelectText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllBySelectText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryBySelectText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
